File: leetcode/1438_Longest_Continuous_Subarray_With_Absolute_Diff_Less_Than_or_Equal_to_Limit.py
import heapq
from typing import List
from collections import deque
from bisect import insort_left, bisect_left


class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        # 2 Pointers & Binary search - Time: O(nlogn), Space: O(n)
        # i, sorted_nums = 0, []
        # for j, x in enumerate(nums):
        #     insort_left(sorted_nums, x)
        #     if sorted_nums[-1] - sorted_nums[0] > limit:
        #         sorted_nums.pop(bisect_left(sorted_nums, nums[i]))
        #         i += 1
        # return j - i + 1

        # 2 Pointers & Heap - Time: O(nlogn), Space: O(n)
        # minq, maxq = [], []
        # i, res = 0, 0
        # for j, x in enumerate(nums):
        #     heapq.heappush(minq, (x, j))
        #     heapq.heappush(maxq, (-x, j))
        #     if -maxq[0][0] - minq[0][0] > limit:
        #         i = min(minq[0][1], maxq[0][1]) + 1
        #         while minq[0][1] < i:
        #             heapq.heappop(minq)
        #         while maxq[0][1] < i:
        #             heapq.heappop(maxq)
        #     res = max(res, j - i + 1)
        # return res

        # 2 Pointers & Sliding Window Max/Min - Time: O(n), Space: O(n)
        mini, maxi = deque(), deque()
        i = 0
        for j, x in enumerate(nums):
            while mini and mini[-1] > x:
                mini.pop()
            mini.append(x)
            while maxi and maxi[-1] < x:
                maxi.pop()
            maxi.append(x)
            if maxi[0] - mini[0] > limit:
                if mini[0] == nums[i]:
                    mini.popleft()
                if maxi[0] == nums[i]:
                    maxi.popleft()
                i += 1
        return len(nums) - i

File: leetcode/14/1438_Longest_Continuous_Subarray_With_Absolute_Diff_Less_Than_or_Equal_to_Limit.java
import java.util.Comparator;
import java.util.PriorityQueue;

class Solution {
    public int longestSubarray(int[] nums, int limit) {
        // Heap, Sliding Window - Time: O(nlogn), Space: O(n)
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a, b) -> b[0] - a[0]);

        int l = 0, res = 0;
        for (int r = 0; r < nums.length; r++) {
            minHeap.offer(new int[] { nums[r], r });
            maxHeap.offer(new int[] { nums[r], r });

            while (maxHeap.peek()[0] - minHeap.peek()[0] > limit) {
                l = Math.min(minHeap.peek()[1], maxHeap.peek()[1]) + 1;
                while (minHeap.peek()[1] < l) {
                    minHeap.poll();
                }
                while (maxHeap.peek()[1] < l) {
                    maxHeap.poll();
                }
            }

            res = Math.max(res, r - l + 1);
        }

        return res;
    }
}
